com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs

```diff
@@ -141,41 +183,45 @@ public override void OnImportAsset(AssetImportContext ctx)
 
                 ReportErrors(graph, shader, path, importLog);
 
-                EditorMaterialUtility.SetShaderDefaults(
-                    shader,
-                    configuredTextures.Where(x => x.modifiable).Select(x => x.name).ToArray(),
-                    configuredTextures.Where(x => x.modifiable).Select(x => EditorUtility.InstanceIDToObject(x.textureId) as Texture).ToArray());
-                EditorMaterialUtility.SetShaderNonModifiableDefaults(
-                    shader,
-                    configuredTextures.Where(x => !x.modifiable).Select(x => x.name).ToArray(),
-                    configuredTextures.Where(x => !x.modifiable).Select(x => EditorUtility.InstanceIDToObject(x.textureId) as Texture).ToArray());
+                SetupShaderTextureDefaults(shader, configuredTextures);
+
+                if (graph.hasTerrainTarget) // TODO: Do this generically, not just for terrain.
+                                            // "take all the subshaders, and group them by ShaderID, then generate one shader per ShaderID it finds"
+                {
+                    // Generate
+                    configuredTextures.Clear();
+                    var basemapGenText = GetShaderText(path, out configuredTextures, assetCollection, graph, shaderIdx: (int)TerrainSubTarget.TerrainShaders.BasemapGen);
+
+                    // Output separate file for debug
+                    string bmgPath = "Assets/BasemapGen.shader";
+                    string fullPath = Path.GetFullPath(bmgPath);
+                    File.WriteAllText(fullPath, basemapGenText);
+                    AssetDatabase.ImportAsset(bmgPath);
+
+                    // Add to shadergraph file as subasset
+                    //Shader basemapGenShader = CreateShader(ctx, basemapGenText, false);
+                    //ReportErrors(graph, basemapGenShader, path);
+                    //SetupShaderTextureDefaults(basemapGenShader, configuredTextures);
+                    //ctx.AddObjectToAsset("Basemap Gen Shader", basemapGenShader);
+
+                    configuredTextures.Clear();
+                    var basePassText = GetShaderText(path, out configuredTextures, assetCollection, graph, shaderIdx: (int)TerrainSubTarget.TerrainShaders.Basemap);
+                    string basePath = "Assets/BasePass.shader";
+                    string fullBasePath = Path.GetFullPath(basePath);
+                    File.WriteAllText(fullBasePath, basePassText);
+                    AssetDatabase.ImportAsset(basePath);
+
+                    configuredTextures.Clear();
+                    var addPassText = GetShaderText(path, out configuredTextures, assetCollection, graph, shaderIdx: (int)TerrainSubTarget.TerrainShaders.Add);
+                    string addPath = "Assets/AddPass.shader";
+                    string fullAddPath = Path.GetFullPath(addPath);
+                    File.WriteAllText(fullAddPath, addPassText);
+                    AssetDatabase.ImportAsset(addPath);
```

Comment:
Currently, and AddPass shader is created even if the main shader does not have an AddPass (e.g. HDRP does not use AddPass). We should make sure GetShaderText returns something reasonable when a shaderIdx is requested that doesn't exist, and we should avoid generating a shader when one isn't needed.